src/api/l_math_mat4.c

```diff
@@ -159,7 +159,10 @@ int l_lovrMat4Set(lua_State* L) {
     m[14] = position[2];
 
     // 1 more arg or 4 numbers: rotation, otherwise scale + rotation
-    if (top == index || ((top - index) == 3 && lua_type(L, top) == LUA_TNUMBER)) {
+    if (
+      (top == index && luax_isquat(L, index))
+      || ((top - index) == 3 && lua_type(L, top) == LUA_TNUMBER)
+    ) {
```

Comment:
I don't like `if`s that are on multiple lines like this.  Maybe just `if (luax_isquat(L, index) || (top - index == 3 && lua_type(L, top) == LUA_TNUMBER))`?